BSP Tutorial (#38)
datahaven-xyz/datahaven-docs: provide-storage/backup-storage-provider/run-a-bsp-node.md

---
title: Run a BSP Node
description: Guide on how to set up, configure, and run a Backup Storage Provider (BSP) node on the DataHaven network using Docker.
---

# Run a BSP Node

This guide explains how to prepare your project workspace, run the node with Docker, select the correct chain spec, and inject your BSP service key. By the end, you’ll have a fully configured BSP that is ready to join the DataHaven network.

## Prerequisites

Before you begin, ensure you have the following:

- macOS or Linux operating system
- Storage Capacity (minimum 1 TB, recommended 2 TB+)
- [Docker](https://www.docker.com/){target=\_blank} and [Docker Compose](https://docs.docker.com/compose/install/){target=\_blank} installed and running
- A BCSV key of scheme ECDSA (a 32 byte private key) for the BSP node's on-chain identity and signing
- Sufficient account balance for deposits and collateral
- Stable network connection
- Open network ports (30333, optionally 9944)

### Hardware Requirements

BSPs have similar hardware requirements to MSPs as they store backup data and must reliably submit proofs of storage.

#### Specifications

> 

Reviewer: ```suggestion
```

I try not to use h4s when possible

| Component | Requirement |
|-----------|-------------|
| **CPU** | 8 physical cores @ 3.4 GHz (Intel Ice Lake+ or AMD Zen3+) |
| **RAM** | 32 GB DDR4 ECC |
| **Storage (System)** | 500 GB NVMe SSD (chain data) |
| **Storage (User Data)** | 1 TB NVMe SSD or HDD (minimum) |
| **Network** | 500 Mbit/s symmetric |

#### Important Considerations

- **Separate storage volumes**: Keep chain data and user data on separate volumes for better I/O performance
- **Storage expansion**: Plan for growth; user data storage should be easily expandable
- **max-storage-capacity**: Set this CLI flag to **80% of available physical disk space** to leave headroom for filesystem overhead and temporary files
- **Cloud compatible**: BSPs can run effectively on cloud VPS with dedicated storage volumes

> - **Proof submission**: Ensure reliable network connectivity for timely proof submissions

Reviewer: These titles are pretty inconsistent, in some we're listing actions or concepts and others we're listing configurations. I'd change this to something like:

```
Use separate storage volumes: Keep chain and user data separate for better I/O performance.

Plan for storage growth: Ensure user data storage is easily expandable.

Limit disk usage: Set `max-storage-capacity` to 80% of available physical space.

Optimize for cloud: Run BSPs on cloud VPS with dedicated storage volumes.

Ensure proof reliability: Maintain stable network connectivity for timely proof submissions.
```

Thank you chatgpt. Also I didn't really read the whole thing to make sure it matched, just the titles to make sure they aligned, so if you intend to use the above, please read it and make sure it makes sense 😆

Reviewer: Note the punctuation at the end of the descriptions though 🙂

### Deposit Requirements

- **Base Deposit**: 100 MOCK (`SpMinDeposit`)
- **Per GiB**: 2 MOCK (`DepositPerData`)

> - **Formula**: 100 + (`capacity_in_gib` × 2) + buffer

Reviewer: This is kind of confusing cause of the format that's being used. Like the stuff in parenthesis for top 2 items is just like this is what it's related to then in the last one it's like oh wait this is not what it's related to it's required for the formula

Examples:

- **800 GiB capacity**: 100 + (800 × 2) = 1,700 MOCK required (1,800 MOCK recommended)
- **1.6 TiB capacity**: 100 + (1,638 × 2) = 3,376 MOCK required (3,500+ MOCK recommended)

The deposit is held (reserved) from your account when you start the BSP registration process and remains held while you operate as a BSP. The deposit is returned when you deregister as a BSP.

!!! note
    Your BSP account must be funded before BSP registration

## Project Structure

This is how the project structure will look like, once everything is set up.

```
datahaven-bsp-node/
├── datahaven-node
├── bsp-data
│   └── chains
│   │   ├── datahaven_stagenet_local
│   │   │   └── keystore
│   │   └── datahaven_testnet
│   │       └── keystore
│   bsp-storage    
│   └── storagehub    
│       ├── file_storage
│       └── forest_storage      
├── Dockerfile
├── docker-compose.yml
└── bsp.log    
```

## Project Setup

> Before running a BSP node, you will need to obtain the `datahaven-node` client binary and the chain specifications for the network you want to join.

Reviewer: Or you could just use one section that has steps like 1. Download the latest client release 2. Download the chain specs

### Download Latest Client Release

From the [Releases](https://github.com/datahaven-xyz/datahaven/releases){target=\_blank} section of the DataHaven repo, the latest version of the [`datahaven-node` binary](https://github.com/datahaven-xyz/datahaven/releases/download/{{ networks.testnet.client_version }}/datahaven-node){target=_\blank} can be found. Currently, the latest version is {{ networks.testnet.client_version }}.

Make sure to download it in the root of your `datahaven-bsp-node` folder.

### Download Desired Chain Specs

Next, [download the testnet chain specs](/downloads/datahaven-testnet-raw-specs.json){target=\_blank} and include them in the root of your project. Make sure the file is called `datahaven-testnet-raw-specs.json`. The specs you use dictate to which DataHaven network your BSP will connect and interact with.

> ## Docker Configuration for the BSP Node

Reviewer: This should be action-oriented; something like "Configure Docker for the BSP Node"

> In this section you’ll create the `Dockerfile`, `docker-compose.yml`, and `.dockerignore` needed to run the node, mount your keystore, load the chain spec, and expose the required ports. After this step, your BSP can run cleanly in an isolated, reproducible container that is easy to start, stop, and upgrade.

Reviewer: ```suggestion
In this section, you’ll create the `Dockerfile`, `docker-compose.yml`, and `.dockerignore` needed to run the node, mount your keystore, load the chain spec, and expose the required ports. After this step, your BSP can run cleanly in an isolated, reproducible container that is easy to start, stop, and upgrade.
```

    # Note: This Dockerfile is architecture-agnostic. Use
    # `docker build --platform=linux/amd64` on Apple Silicon
    # so Rosetta can emulate x86-64 binaries.
    FROM ubuntu:22.04

    # Install runtime dependencies for datahaven-node
    RUN apt-get update && apt-get install -y \
        libpq5 \
        ca-certificates \
    && rm -rf /var/lib/apt/lists/*

    # Copy the datahaven-node binary into the image
    COPY datahaven-node /usr/local/bin/datahaven-node
    RUN chmod +x /usr/local/bin/datahaven-node

    # Default entrypoint: just run the node
    ENTRYPOINT ["datahaven-node"]
    ```

    This `Dockerfile` builds a minimal Ubuntu-based image containing the `datahaven-node` binary and its required runtime libraries. It installs only the necessary dependencies, copies the node executable into the container, and marks it as runnable. On macOS with Apple Silicon, the image must be built with `--platform=linux/amd64` so Docker can emulate the `x86-64` Linux environment that the precompiled DataHaven binaries expect. The default entrypoint runs `datahaven-node` directly, allowing the container to behave like a fully configured BSP node.

> 3. Create a `.dockerignore` file

Reviewer: ```suggestion
3. Create a `.dockerignore` file.
```

> 4. Run the following command in the terminal to list the `data` folder within the `.dockerignore` file:

Reviewer: ```suggestion
4. List the `data` folder within the `.dockerignore` file:
```

let's just get rid of the extra words

    ```bash
    echo "data" > .dockerignore
    ```

    This way the `data` folder and its contents won't get included in the docker image.

> 5. Build the docker image with the following command:

Reviewer: ```suggestion
5. Build the Docker image:
```

can you pls check to make sure Docker is capitalized everywhere else?

>     !!! note

Reviewer: ```suggestion
    !!! note "Note for macOS (Apple Silicon) users"
```

    ```bash
    docker build --platform=linux/amd64 -t datahaven-bsp:latest .
    ```

    The image name can be anything. In this tutorial, we'll use `datahaven-bsp:latest`. 
    
    You won't run the BSP this way, but if you wanted to, right now you could run a BSP in the background through a `datahaven-bsp` docker container with a command like this:

    !!! note
        Include `--platform=linux/amd64` only if using Apple Silicon.

    ```bash
    docker run -d \
      --platform=linux/amd64 \
      --name datahaven-bsp \
      --restart unless-stopped \
      -v "./data":/data \
      datahaven-bsp:latest \
        --provider \
        --provider-type bsp \
        --max-storage-capacity 10737418240 \
        --jump-capacity 1073741824 \
        --storage-layer rocks-db \
        --storage-path /data
    ```

>     To display this container's logs in terminal, run:

Reviewer: Is this step 6, or still part of a really long step 5? EDIT: oh wow, it's still step 5. If it's optional stuff maybe put it in an expandable?

Reviewer: if it's not optional, we're now running the node instead of building it so i'd separate it out into it's own step

    ```bash
    docker logs -f datahaven-bsp
    ```

    !!! note
        To stop displaying the logs in terminal run `Ctrl+C`, but keep in mind that it won't actually stop the container from running.

    Once the `datahaven-bsp` container is running, these commands are useful to keep in mind:

    ```bash
    # stop the container
    docker stop datahaven-bsp 

    # start the container
    docker start datahaven-bsp 

    # restart the container
    docker restart datahaven-bsp

    # delete the stopped container entirely
    docker rm datahaven-bsp 
    ```

    The logs output should be something like:

    --8<-- 'code/provide-data/backup-storage-provider/end-to-end-bsp-onboarding/output-01.html'

    If you have spun up this `datahaven-bsp` container, prior to the upcoming step, make sure to stop it and remove it:

    ```bash
    docker stop datahaven-bsp
    docker rm datahaven-bsp
    ```

> 6. In the root of your project create a `docker-compose.yml` file

Reviewer: ```suggestion
6. In the root of your project create a `docker-compose.yml` file.
```

    ```bash
    docker compose run --rm datahaven-bsp \
    key inspect --scheme ecdsa --output-type json "$SEED" | jq -r '.ss58PublicKey'
    ```

    The output should look something like this:

    --8<-- 'code/provide-data/backup-storage-provider/end-to-end-bsp-onboarding/output-04.html'

### Insert BCSV Key (ECDSA)

1. Make sure to stop and remove the container by running:

    ```bash
    docker compose down
    ```

2. Run the following command:

    !!! note
        The `--base-path` flag is crucial for keeping your keys across restarts.

    ```bash
    docker compose run --rm \
    datahaven-bsp key insert \
      --base-path /data \
      --chain /testnet-chain-spec.json \
      --key-type bcsv \
      --scheme ecdsa \
      --suri "$SEED"
    ```

    This command writes the resulting key into `/bsp-data/chains/datahaven_testnet/keystore` on your host. You can check if the command was successful by running:

    ```bash
    find bsp-data -maxdepth 4 -type d -name keystore
    ls bsp-data/chains/*/keystore
    ```

    The output should look something like this:

    --8<-- 'code/provide-data/backup-storage-provider/end-to-end-bsp-onboarding/output-05.html'

> Finally, you have a running BSP node within an easily maintainable docker container and an injected keystore on the DataHaven network you've specified.

Reviewer: ```suggestion
Finally, you have a running BSP node in an easily maintainable Docker container, with an injected keystore on the DataHaven network you've specified.
```

## Next Steps

<div class="grid cards" markdown>

-  <a href="/provide-storage/backup-storage-provider/verify-bsp-on-chain" markdown>:material-arrow-right: 

    **Create A Bucket**

    Follow this guide to verify your BSP on-chain and thus enable your BSP node to participate in replicating DataHaven's files.

    </a>

> -  <a href="/provide-storage/backup-storage-provider/end-to-end-bsp-onboarding.md" markdown>:material-arrow-right:

Reviewer 2: ```suggestion
-  <a href="/provide-storage/backup-storage-provider/end-to-end-bsp-onboarding/" markdown>:material-arrow-right:
```

8. Run the BSP:

    ```bash
    docker compose up -d
    ```

    The output will look something like this:

    --8<-- 'code/provide-storage/backup-storage-provider/end-to-end-bsp-onboarding/output-03.html'

9. Check BSP's logs:

    ```bash
    docker compose logs -f | tee bsp.log
    ```

    Make sure the `Chain specification` log is displaying the network that matches the chain spec you are using:

    --8<-- 'code/provide-storage/backup-storage-provider/end-to-end-bsp-onboarding/output-04.html'

### Useful Docker Commands

A collection of helpful Docker Compose commands you’ll use while developing or debugging your BSP node:

    ```bash
    # run the container in the background
    docker compose up -d

    # stop and remove the container
    docker compose down

    # check status and logs while the container is running
    docker compose ps

    # continuously tail logs in terminal
    docker compose logs -f

    # continuously display logs in terminal and save them into a file
    docker compose logs -f | tee bsp.log

    # continuously stream and continuously save all logs into a file
    docker compose logs -f > bsp.log

>     ```

Reviewer: The indent here is causing problems when this renders. But let's break these out to individual commands so that users can copy which ones they need
